server/entity/splash_potion.go

```diff
@@ -0,0 +1,239 @@
+
+	s.ticksLived++
+	m.Send()
+
+	if m.pos[1] < cube.MinY && current%10 == 0 {
+		s.closeNextTick = true
+		return
+	}
+
+	if result != nil {
+		w := s.World()
+		aabb := s.AABB().Translate(m.pos)
+
+		effects := s.t.Effects()
+		hasEffects := len(effects) > 0
+
+		colour := color.RGBA{R: 0x38, G: 0x5d, B: 0xc6, A: 0xff}
+		if hasEffects {
+			colour, _ = effect.ResultingColour(effects)
+		}
+
+		w.AddParticle(m.pos, particle.Splash{Colour: colour})
+		w.PlaySound(m.pos, sound.GlassBreak{})
+
+		if hasEffects {
+			ignore := func(entity world.Entity) bool {
+				_, canSplash := entity.(splashable)
+				return !canSplash || entity == s
+			}
+
+			for _, otherEntity := range w.EntitiesWithin(aabb.GrowVec3(mgl64.Vec3{4.125, 2.125, 4.125}), ignore) {
+				splashEntity := otherEntity.(splashable)
+
+				distance := world.Distance(EyePosition(splashEntity), m.pos)
+				if distance > 4 {
+					continue
+				}
+
+				distanceMultiplier := 1 - (distance / 4)
+				if entityResult, ok := result.(trace.EntityResult); ok && entityResult.Entity() == otherEntity {
+					distanceMultiplier = 1.0
+				}
+
+				for _, eff := range effects {
+					if _, ok := eff.Type().(effect.LastingType); !ok {
+						splashEntity.AddEffect(eff.WithPotency(distanceMultiplier))
+						continue
+					}
+
+					splashEntity.AddEffect(eff.WithDuration(time.Duration(float64(eff.Duration().Milliseconds())*0.75*distanceMultiplier) * time.Millisecond))
+				}
+			}
+		} else if blockResult, ok := result.(trace.BlockResult); ok && s.t.Equals(potion.Water()) {
+			blockPos := blockResult.BlockPosition().Side(blockResult.Face())
+			if w.Block(blockPos) == fire() {
+				w.SetBlock(blockPos, air())
+			}
+
+			for _, f := range cube.HorizontalFaces() {
+				horizontalPos := blockPos.Side(f)
+				if w.Block(horizontalPos) == fire() {
+					w.SetBlock(horizontalPos, air())
+				}
+			}
+		}
+
+		s.closeNextTick = true
+	}
+}
+
+// ignores returns whether the SplashPotion should ignore collision with the entity passed.
+func (s *SplashPotion) ignores(entity world.Entity) bool {
+	_, ok := entity.(Living)
+	return !ok || entity == s || (s.ticksLived < 5 && entity == s.owner)
+}
+
+// New creates a SplashPotion with the position, velocity, yaw, and pitch provided. It doesn't spawn the SplashPotion,
+// only returns it.
+func (s *SplashPotion) New(pos, vel mgl64.Vec3, yaw, pitch float64) world.Entity {
+	splash := NewSplashPotion(pos, yaw, pitch, nil)
+	splash.vel = vel
+	return splash
+}
+
+// Owner ...
+func (s *SplashPotion) Owner() world.Entity {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	return s.owner
+}
+
+// Own ...
+func (s *SplashPotion) Own(owner world.Entity) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.owner = owner
+}
+
+// DecodeNBT decodes the properties in a map to a SplashPotion and returns a new SplashPotion entity.
+func (s *SplashPotion) DecodeNBT(data map[string]interface{}) interface{} {
+	p := s.New(
+		nbtconv.MapVec3(data, "Pos"),
+		nbtconv.MapVec3(data, "Motion"),
```

Comment:
After you add the potion type to the constructor, you can just use that instead of this, it'll be cleaner

Reply:
I can't change the constructor of New without causing conflicts to other implementations of New. Do you want me to use NewSplashPotion instead? If so, I'll still have to have another line after it anyways, to set the velocity, since NewSplashPotion doesn't allow for it.

Reply:
Is that okay?

Reply:
Also, by changing the constructor of NewSplashPotion, New will somehow need to set a default potion otherwise that won't work either. If I should continue this way, shall I just make it set it to water?

Reply:
I was referring to NewSplashPotion, not .New. NewSplashPotion should have the potion type anyways for consistency.

Reply:
New does use NewSplashPotion though, so I will have to initialize it with some potion anyways. I assume this should just be Water?